tests/unit/test_asymencryption.cpp

```diff
@@ -0,0 +1,958 @@
+/*
+ * #%L
+ * %%
+ * Copyright (C) 2018 BMW Car IT GmbH
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+#include <algorithm>
+#include <fstream>
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include "IOUtils.h"
+
+#include "asymmetric_encryption.cpp"
+
+using namespace std::string_literals;
+
+using namespace mococrw;
+
+
+
+
+class AsymmetricEncryptionTest : public ::testing::Test
+{
+public:
+    /// \brief Structure to hold a private/public key pair
+    struct KeyPair
+    {
+        std::string _publicKey;
+        std::string _privateKey;
+    };
+
+    /// \brief Structure to hold the nominal data set
+    struct NominalDataSet
+    {
+        // inputs
+        std::string _message;
+        std::vector<uint8_t> _encrypted;
+        std::string _publicKey;
+        std::string _privateKey;
+        openssl::RSAPaddingMode _paddingMode;
+        openssl::DigestTypes _hashing;
+        openssl::DigestTypes _masking;
+        std::vector<uint8_t> _label;
+    };
+
+    /// \brief Structure to hold the data set used to test the message size
+    struct MessageSizeDataSet
+    {
+        // inputs
+        std::string _message;
+        std::string _publicKey;
+        std::string _privateKey;
+        openssl::RSAPaddingMode _paddingMode;
+        openssl::DigestTypes _hashing;
+        openssl::DigestTypes _masking;
+        std::vector<uint8_t> _label;
+        // expected outputs
+        bool _expectThrow;
+    };
+
+protected:
+    /// \brief Creates a shared pointer for a RSA padding based on the given inputs. Simplifies the
+    /// tests by taking advantage of polymorphism
+    std::shared_ptr<RSAPadding> createPadding(const openssl::RSAPaddingMode &mode,
+                                              const openssl::DigestTypes &hashing,
+                                              const openssl::DigestTypes &masking,
+                                              const std::vector<uint8_t> &label);
+
+    static const std::vector<NominalDataSet> nominalDataSet;
+    static const std::vector<MessageSizeDataSet> messageSizeDataSet;
+    static const std::vector<NominalDataSet> encryptionInvalidParametersDataSet;
+
+    static const KeyPair keyPairs1024bit;
+    static const KeyPair keyPairs2048bit;
+};
+
+std::shared_ptr<RSAPadding>
+AsymmetricEncryptionTest::createPadding(const openssl::RSAPaddingMode &mode,
```

Comment:
Couldn't this function be useful elsewhere? Perhaps it could be included in the _padding_mode.h_ file.